docs/proposals/ingester-unregister-flexibility.md

```diff
@@ -0,0 +1,129 @@
+
+### Ingester ring membership
+
+By default, an ingester will leave the ingester ring when it is shut down
+(controlled by the [unregister_on_shutdown](https://grafana.com/docs/mimir/v2.11.x/references/configuration-parameters/#ingester) configuration option).
+
+Ingesters leaving the ring on shutdown cause churn during rollouts as series get re-sharded across all ingesters
+(see [grafana/helm-charts#1313](https://github.com/grafana/helm-charts/issues/1313)). This proliferation of series leads
+to excessive ingester CPU usage, and ultimately results in unstable rollouts.
+
+Setting `unregister_on_shutdown` to `false` mitigates the rollout issue, but introduces a new one: multi-zone
+(assuming Mimir runs [zone-aware](https://grafana.com/docs/mimir/v2.11.x/configure/configure-zone-aware-replication/))
+evictions now jeopardise the availability of Mimir.
+
+Mimir needs quorum in order to serve reads and writes. With the current implementation, a single non-`ACTIVE` ingester
+is enough for that ingester's entire zone to be considered unhealthy. If two (assuming a replication factor of three)
+zones are unhealthy, Mimir's ability to serve reads and writes is compromised.
+
+The cost-conscious Mimir operator is now stuck between a rock and a hard place:
+
+|                                               | Single-zone eviction | Multi-zone eviction                                                                          | Rollout                                                                                                                                                                            |
+|-----------------------------------------------|----------------------|----------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `-ingester.ring.unregister-on-shutdown=false` | No problem.          | Guaranteed full loss of read path for all series. Loss of write path for overlapping series. | No problem. If an eviction happens in another zone than the one currently being rolled out, then we will see the same issues as a multi-zone eviction.                             |
+| `-ingester.ring.unregister-on-shutdown=true`  | No problem.          | Temporary loss of data for overlapping series. No loss of write path.                        | Series churn and ingester CPU problems. If an eviction happens in another zone than the one currently being rolled out, then we will see the same issues as a multi-zone eviction. |
+
+## Goals
+
+* Provide a way to dynamically (i.e., _without_ re-starting the ingester) set whether an ingester should unregister or not on next shutdown.
+* Give Mimir operators more flexibility to choose between availability and consistency.
+
+## Proposal
+
+A new `/ingester/prepare-unregister` HTTP endpoint is added.
```

Comment:
Nit: I'd suggest to call it this endpoint similar to the option, ie. `/ingester/unregister-on-shutdown`.

Reply:
Good idea.

Action taken: renamed `/ingester/prepare-unregister` to `/ingester/unregister-on-shutdown`.